### PR TITLE
Add E2E job which build and deploy operators

### PR DESCRIPTION
**Pull request by bshewale**

### CODE REVIEW

**Comment by gibizer** [comment]
The deploy job doesn't seem to work:

```
[36mINFO[0m[2023-02-17T15:55:36Z] Running step nova-operator-build-deploy-openstack-k8s-operators-build. 
[36mINFO[0m[2023-02-17T16:25:46Z] pod pending timeout: container "test" has not started in 30m0s: 
* Container test is not ready with reason ImagePullBackOff and message Back-off pulling image "image-registry.openshift-image-registry.svc:5000/ci-op-cgzjrtyl/stable-latest-nova-operator-build-deploy:openstack-operator-ci-build-deploy"
```
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/36504/rehearse-36504-pull-ci-openstack-k8s-operators-nova-operator-master-nova-operator-build-deploy/1626610567751929856/build-log.txt

**Comment by Sandeepyadav93** [comment]
/ph-rehearse pull-ci-openstack-k8s-operators-nova-operator-master-nova-operator-build-deploy

File: ci-operator/config/openstack-k8s-operators/nova-operator/openstack-k8s-operators-nova-operator-master.yaml [hunk 1]
```diff
@@ -1,14 +1,18 @@
...
+    name: opm-builder
+    namespace: ci
+    tag: latest
 build_root:
-  image_stream_tag:
-    name: release
-    namespace: openshift
-    tag: rhel-8-release-golang-1.18-openshift-4.12
+  from_repository: true
 images:
```
Comment (Sandeepyadav93) [inline ci-operator/config/openstack-k8s-operators/nova-operator/openstack-k8s-operators-nova-operator-master.yaml:16]: Add below under images to do the image build:

~~~
- dockerfile_literal: |
    FROM src
    RUN pip3 install python-openstackclient
    COPY oc /usr/bin/oc
    COPY opm /usr/bin/opm
  from: src
  inputs:
    cli:
      paths:
      - destination_dir: .
        source_path: /usr/bin/oc
    opm-builder:
      paths:
      - destination_dir: .
        source_path: /bin/opm
  to: openstack-operator-ci-build-deploy
~~~

**Review by Sandeepyadav93 — CHANGES_REQUESTED** [review]

File: ci-operator/config/openstack-k8s-operators/nova-operator/openstack-k8s-operators-nova-operator-master.yaml (hunk 1 above)
Comment (bshewale) [reply ci-operator/config/openstack-k8s-operators/nova-operator/openstack-k8s-operators-nova-operator-master.yaml:16]: Done

**Comment by gibizer** [comment]
/pj-rehearse pull-ci-openstack-k8s-operators-nova-operator-master-nova-operator-build-deploy

**Comment by gibizer** [comment]
I don't know why the bundle build fails in CI
```
/alabama/nova-operator/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
Error: err: exit status 1: stderr: go: downloading github.com/evanphx/json-patch v5.6.0+incompatible
go: github.com/emicklei/go-restful/v3@v3.10.1: missing go.sum entry; to add it:
	go mod download github.com/emicklei/go-restful/v3
```
I cannot reproduce this with locally running:
```
+ GOWORK=
+ make build bundle
```

**Comment by viroel** [comment]
@bshewale can you pls rebase this PR? thanks

**Review by viroel — CHANGES_REQUESTED** [review]
Thanks Bhagyashri, few comments added

**Comment by Sandeepyadav93** [comment]
/pj-rehearse pull-ci-openstack-k8s-operators-nova-operator-master-nova-operator-build-deploy

**Comment by Sandeepyadav93** [comment]
I was able to repliacte the issue by pulling container we use in CI to run the job

~~~
]$ podman run -it registry.build02.ci.openshift.org/ci-op-xwwvy412/pipeline:src bash
[root@3e87062cd025 nova-operator]# unset GOFLAGS
[root@3e87062cd025 nova-operator]# GOWORK='' make build
test -s /go/src/github.com/openstack-k8s-operators/nova-operator/bin/controller-gen || GOBIN=/go/src/github.com/openstack-k8s-operators/nova-operator/bin go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.10.0
go: downloading sigs.k8s.io/controller-tools v0.10.0
go: downloading github.com/spf13/cobra v1.4.0
go: downloading gopkg.in/yaml.v2 v2.4.0
go: downloading golang.org/x/tools v0.1.12
go: downloading github.com/fatih/color v1.12.0
go: downloading k8s.io/api v0.25.0
go: downloading k8s.io/apimachinery v0.25.0
go: downloading github.com/gobuffalo/flect v0.2.5
go: downloading k8s.io/apiextensions-apiserver v0.25.0
go: downloading gopkg.in/yaml.v3 v3.0.1
go: downloading sigs.k8s.io/yaml v1.3.0
go: downloading github.com/spf13/pflag v1.0.5
go: downloading github.com/mattn/go-colorable v0.1.8
go: downloading github.com/mattn/go-isatty v0.0.12
go: downloading github.com/gogo/protobuf v1.3.2
go: downloading k8s.io/utils v0.0.0-20220728103510-ee6ede2d64ed
go: downloading github.com/google/gofuzz v1.1.0
go: downloading k8s.io/klog/v2 v2.70.1
go: downloading sigs.k8s.io/structured-merge-diff/v4 v4.2.3
go: downloading sigs.k8s.io/json v0.0.0-20220713155537-f223a00ba0e2
go: downloading gopkg.in/inf.v0 v0.9.1
go: downloading golang.org/x/sys v0.0.0-20220722155257-8c9f86f7a55f
go: downloading golang.org/x/mod v0.6.0-dev.0.20220419223038-86c51ed26bb4
go: downloading golang.org/x/net v0.0.0-20220722155237-a158d28d115b
go: downloading github.com/go-logr/logr v1.2.3
go: downloading github.com/json-iterator/go v1.1.12
go: downloading github.com/modern-go/reflect2 v1.0.2
go: downloading github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd
go: downloading golang.org/x/text v0.3.7
/go/src/github.com/openstack-k8s-operators/nova-operator/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
Error: err: exit status 1: stderr: go: downloading github.com/evanphx/json-patch v5.6.0+incompatible
go: github.com/golang/protobuf@v1.5.3: missing go.sum entry; to add it:
	go mod download github.com/golang/protobuf

Usage:
  controller-gen [flags]

Examples:
	# Generate RBAC manifests and crds for all types under apis/,
	# outputting crds to /tmp/crds and everything else to stdout
	controller-gen rbac:roleName=<role name> crd paths=./apis/... output:crd:dir=/tmp/crds output:stdout

	# Generate deepcopy/runtime.Object implementations for a particular file
	controller-gen object paths=./apis/v1beta1/some_types.go

	# Generate OpenAPI v3 schemas for API packages and merge them into existing CRD manifests
	controller-gen schemapatch:manifests=./manifests output:dir=./manifests paths=./pkg/apis/... 

	# Run all the generators for a given project
	controller-gen paths=./apis/...

	# Explain the markers for generating CRDs, and their arguments
	controller-gen crd -ww


Flags:
  -h, --detailed-help count   print out more detailed help
                              (up to -hhh for the most detailed output, or -hhhh for json output)
      --help                  print out usage and a summary of options
      --version               show version
  -w, --which-markers count   print out all markers available with the requested generators
                              (up to -www for the most detailed output, or -wwww for json output)


Options


generators

+webhook                                                                                                                                           package  generates (partial) {Mutating,Validating}WebhookConfiguration objects.                        
+schemapatch[:generateEmbeddedObjectMeta=<bool>],manifests=<string>[,maxDescLen=<int>]                                                             package  patches existing CRDs with new schemata.                                                      
+rbac:roleName=<string>                                                                                                                            package  generates ClusterRole objects.                                                                
+object[:headerFile=<string>][,year=<string>]                                                                                                      package  generates code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.  
+crd[:allowDangerousTypes=<bool>][,crdVersions=<[]string>][,generateEmbeddedObjectMeta=<bool>][,ignoreUnexportedFields=<bool>][,maxDescLen=<int>]  package  generates CustomResourceDefinition objects.                                                   


generic

+paths=<[]string>  package  represents paths and go-style path patterns to use as package roots.   


output rules (optionally as output:<generator>:...)

+output:artifacts[:code=<string>],config=<string>  package  outputs artifacts to different locations, depending on whether they're package-associated or not.   
+output:dir=<string>                               package  outputs each artifact to the given directory, regardless of if it's package-associated or not.      
+output:none                                       package  skips outputting anything.                                                                          
+output:stdout                                     package  outputs everything to standard-out, with no separation.                                             

run `controller-gen object:headerFile=hack/boilerplate.go.txt paths=./... -w` to see all available markers, or `controller-gen object:headerFile=hack/boilerplate.go.txt paths=./... -h` for usage
make: *** [Makefile:113: generate] Error 1
~~~

Second attempt:

~~~
[root@3e87062cd025 nova-operator]# GOWORK='' make build
test -s /go/src/github.com/openstack-k8s-operators/nova-operator/bin/controller-gen || GOBIN=/go/src/github.com/openstack-k8s-operators/nova-operator/bin go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.10.0
/go/src/github.com/openstack-k8s-operators/nova-operator/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
Error: err: exit status 1: stderr: go: github.com/golang/protobuf@v1.5.3: missing go.sum entry; to add it:
	go mod download github.com/golang/protobuf

Usage:
  controller-gen [flags]

Examples:
	# Generate RBAC manifests and crds for all types under apis/,
	# outputting crds to /tmp/crds and everything else to stdout
	controller-gen rbac:roleName=<role name> crd paths=./apis/... output:crd:dir=/tmp/crds output:stdout

	# Generate deepcopy/runtime.Object implementations for a particular file
	controller-gen object paths=./apis/v1beta1/some_types.go

	# Generate OpenAPI v3 schemas for API packages and merge them into existing CRD manifests
	controller-gen schemapatch:manifests=./manifests output:dir=./manifests paths=./pkg/apis/... 

	# Run all the generators for a given project
	controller-gen paths=./apis/...

	# Explain the markers for generating CRDs, and their arguments
	controller-gen crd -ww


Flags:
  -h, --detailed-help count   print out more detailed help
                              (up to -hhh for the most detailed output, or -hhhh for json output)
      --help                  print out usage and a summary of options
      --version               show version
  -w, --which-markers count   print out all markers available with the requested generators
                              (up to -www for the most detailed output, or -wwww for json output)


Options


generators

+webhook                                                                                                                                           package  generates (partial) {Mutating,Validating}WebhookConfiguration objects.                        
+schemapatch[:generateEmbeddedObjectMeta=<bool>],manifests=<string>[,maxDescLen=<int>]                                                             package  patches existing CRDs with new schemata.                                                      
+rbac:roleName=<string>                                                                                                                            package  generates ClusterRole objects.                                                                
+object[:headerFile=<string>][,year=<string>]                                                                                                      package  generates code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.  
+crd[:allowDangerousTypes=<bool>][,crdVersions=<[]string>][,generateEmbeddedObjectMeta=<bool>][,ignoreUnexportedFields=<bool>][,maxDescLen=<int>]  package  generates CustomResourceDefinition objects.                                                   


generic

+paths=<[]string>  package  represents paths and go-style path patterns to use as package roots.   


output rules (optionally as output:<generator>:...)

+output:artifacts[:code=<string>],config=<string>  package  outputs artifacts to different locations, depending on whether they're package-associated or not.   
+output:dir=<string>                               package  outputs each artifact to the given directory, regardless of if it's package-associated or not.      
+output:none                                       package  skips outputting anything.                                                                          
+output:stdout                                     package  outputs everything to standard-out, with no separation.                                             

run `controller-gen object:headerFile=hack/boilerplate.go.txt paths=./... -w` to see all available markers, or `controller-gen object:headerFile=hack/boilerplate.go.txt paths=./... -h` for usage
make: *** [Makefile:113: generate] Error 1
~~~

Adding below in api/go.sum solved the issue:-

~~~
[sandeepyadav@localhost tmp]$ diff /tmp/nova-operator/api/go.sum go.sum.modified 
137a138
> github.com/golang/protobuf v1.5.3/go.mod h1:XVQd3VNwM+JqD3oG2Ue2ip4fOMUkwXdXDdiuN0vRsmY=
222a224
> github.com/onsi/ginkgo/v2 v2.9.1/go.mod h1:FEcmzVcCHl+4o9bQZVab+4dC9+j+91t2FHSzmGAPfuo=
223a226
> github.com/onsi/gomega v1.27.3/go.mod h1:5vG284IBtfDAmDyrK+eGyZmUgUlmi+Wngqo557cZ6Gw=
372a376
> golang.org/x/net v0.8.0/go.mod h1:QVkue5JL9kW//ek3r6jTKnTFis1tRmNAW2P1shuFdJc=
434a439
> golang.org/x/sys v0.6.0/go.mod h1:oPkhp1MJrh7nUepCBck5+mAzfO9JrbApNNgaTdGDITg=
437a443
> golang.org/x/term v0.6.0/go.mod h1:m6U89DPEgQRMq3DNkDClhWw02AUbt2daBVO4cn4Hv9U=
446a453
> golang.org/x/text v0.8.0/go.mod h1:e1OnstbJyHTd6l/uOt8jFFHp6TRDWZR/bV3emEE/zU8=
623a631
> k8s.io/client-go v0.26.2/go.mod h1:u5EjOuSyBa09yqqyY7m3abZeovO/7D/WehVVlZ2qcqU=
~~~~

Passed this time:

~~~
[root@3e87062cd025 nova-operator]# GOWORK='' make build
test -s /go/src/github.com/openstack-k8s-operators/nova-operator/bin/controller-gen || GOBIN=/go/src/github.com/openstack-k8s-operators/nova-operator/bin go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.10.0
/go/src/github.com/openstack-k8s-operators/nova-operator/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
test -f go.work || go work init
go work use .
go work use ./api
go work sync
go vet ./...
go vet ./api/...
go build -ldflags "-X "main.gitBuildHash=cnosp-415-g4205828-dirty"" -o bin/manager main.go
[root@3e87062cd025 nova-operator]# echo $?
0
~~~~

@gibizer Could you please check why the below entry is missing in api/go.sum (its present in go.sum in main dir)

~~~
github.com/golang/protobuf v1.5.3/go.mod h1:XVQd3VNwM+JqD3oG2Ue2ip4fOMUkwXdXDdiuN0vRsmY=
~~~

https://github.com/openstack-k8s-operators/nova-operator/blob/master/api/go.sum#L137
~~~
github.com/golang/protobuf v1.5.3 h1:KhyjKVUg7Usr/dYsdSqoFveMYd5ko72D+zANwlG1mmg=
github.com/google/btree v0.0.0-20180813153112-4030bb1f1f0c/go.mod h1:lNA+9X1NB3Zf8V7Ke586lFgjr2dZNuvo3lPJSGZ5JPQ=
~~~
https://github.com/openstack-k8s-operators/nova-operator/blob/master/go.sum#L142
~~~
github.com/golang/protobuf v1.5.3 h1:KhyjKVUg7Usr/dYsdSqoFveMYd5ko72D+zANwlG1mmg=
github.com/golang/protobuf v1.5.3/go.mod h1:XVQd3VNwM+JqD3oG2Ue2ip4fOMUkwXdXDdiuN0vRsmY=
github.com/google/btree v0.0.0-20180813153112-4030bb1f1f0c/go.mod h1:lNA+9X1NB3Zf8V7Ke586lFgjr2dZNuvo3lPJSGZ5JPQ=
~~~

**Comment by Sandeepyadav93** [comment]
/pj-rehearse ack


deploy job is green:

https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/36504/rehearse-36504-pull-ci-openstack-k8s-operators-nova-operator-master-nova-operator-build-deploy/1636670702062931968/artifacts/nova-operator-build-deploy/openstack-k8s-operators-deploy/build-log.txt

**Review by gibizer — APPROVED** [review]
Thank you @bshewale @viroel and @Sandeepyadav93 to made this green.

**Pull request merged**